### PR TITLE
Add the ability to use an alternative API server to perform requests against the cluster.

### PR DESCRIPTION
**Pull request by metlos**
### What does this PR do?
This has some far-reaching consequences.

Because we're now using the identity of the caller of the HTTP request to
perform the calls into the cluster, we need to make sure that the user is
properly authenticated and that we don't escalate the permissions allowed
for the user by using the service account of the OAuth service.

This in turn requires us to reintroduce to somehow remember the identity
of the user during the OAuth flow. This is now done such that each OAuth
flow is assigned a unique key that is stored in the OAuth state and in the
session of the OAuth service (the session id is stored as a cookie). The
bearer token is kept in the session in-memory on the server side. Upon
completing the OAuth flow the key from the OAuth state is used to look up
the bearer token in the session. The OAuth flow fails if the session
doesn't contain the bearer token under the key.

This way we can make sure that the user that started the OAuth flow is the
same as the one who finishes it (by relying on the session).

### What issues does this PR fix or reference?
https://issues.redhat.com/browse/SVPI-81

### How to test this PR?
1. Set up OAuth application in Github
1. Define the following env vars to use the images built from the PRs: `SPIS_IMG=quay.io/lkrejci/service-provider-integration-oauth:svpi-62` and `SPIO_IMG=quay.io/lkrejci/service-provider-integration-operator:svpi-62`
1. Run the `hack/multi-cluster/minikube.sh` script from https://github.com/redhat-appstudio/service-provider-integration-operator/pull/106
1. Update the configuration of the OAuth app in Github with the callback URL as instructed by the above script.
1. Connect to the `minikube-a` cluster: `kubectl config use-context minikube-a`
1. Give the default SA permissions - in **operator** repo: `kubectl apply -f hack/give-default-sa-perms-for-accesstokens.yaml`
1. Open the `hack/oauth-ui.html` in browser.
1. Use `hack/get-default-sa-token.sh` to get the k8s token of the default SA and paste it to the UI
1. Create an `SPIAccessTokenBinding` (just apply some from the samples in the operator repo)
1. Get the OAuth URL from the binding (`kubectl get spiaccesstokenbinding -o jsonpath='{.items[0].status.oAuthUrl}'`) and copy it into the UI
1. Click "Submit" in the UI and go through the OAuth flow.

### CODE REVIEW

File: go.mod [hunk 1]
```diff
@@ -215,7 +195,8 @@ require (
 	k8s.io/component-base v0.22.4 // indirect
 	k8s.io/klog/v2 v2.9.0 // indirect
 	k8s.io/kube-openapi v0.0.0-20211109043538-20434351676c // indirect
-	sigs.k8s.io/apiserver-network-proxy/konnectivity-client v0.0.22 // indirect
 	sigs.k8s.io/structured-merge-diff/v4 v4.1.2 // indirect
 	sigs.k8s.io/yaml v1.2.0 // indirect
 )
+
+replace github.com/redhat-appstudio/service-provider-integration-operator => github.com/metlos/service-provider-integration-operator v0.0.0-20220331145451-87195912cff4
```
Comment (sparkoo) [inline go.mod:202]: please don't forget to revert and update operator dep, one PR is merged.

File: go.mod [hunk 2]
```diff
@@ -3,18 +3,19 @@ module github.com/redhat-appstudio/service-provider-integration-oauth
 go 1.17
 
 require (
+	github.com/alexedwards/scs v1.4.1
```
Comment (sparkoo) [inline go.mod:6]: is this completely new dependency? Do we know it's stable and trustworthy?

Comment (sparkoo) [inline go.mod:6]: is this completely new dependency? Do we know it's stable and trustworthy?

File: controllers/controller.go [hunk 3]
```diff
@@ -55,30 +49,7 @@ const (
...
-	cl, err := client.New(kubeConfig, client.Options{Scheme: scheme})
-	if err != nil {
-		return nil, err
-	}
-
-	vaultStorage, err := tokenstorage.NewVaultStorage("spi-oauth", fullConfig.VaultHost, fullConfig.ServiceAccountTokenFilePath)
-	if err != nil {
-		return nil, err
-	}
+func FromConfiguration(fullConfig config.Configuration, spConfig config.ServiceProviderConfiguration, sessionManager *scs.Manager, cl AuthenticatingClient, vaultStorage tokenstorage.TokenStorage, redirectTemplate *template.Template) (Controller, error) {
```
Comment (sparkoo) [inline controllers/controller.go:52]: please rename `vaultStorage` here. It should not depend on implementation, right? It can be any storage.

**Review by sparkoo — APPROVED** [review]
do we have token upload API documented anywhere? Is it worth to put it to README.md here or in spi-operator readme? Or maybe prepare some sample script?

File: go.mod (hunk 2 above)
Comment (metlos) [reply go.mod:6]: It is a new dependency. It is used by some higher profile projects like https://github.com/solo-io/gloo and itself has quite a lot of stars on github. Didn't do any code analysis on it though,

File: controllers/controller.go (hunk 3 above)
Comment (metlos) [reply controllers/controller.go]: 👍🏼

**Comment by metlos** [comment]
> do we have token upload API documented anywhere? Is it worth to put it to README.md here or in spi-operator readme? Or maybe prepare some sample script?

Good point, I will update the token upload PR with some updates to the README

**Review by sparkoo — APPROVED** [review]

**Pull request merged**